### PR TITLE
reorganize partitions with bindvar support

### PR DESCRIPTION
**Pull request by acharis**
@sougou @bbeaudreault

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -499,22 +499,25 @@ func (node *Set) WalkSubtree(visit Visit) error {
...
-	Table     TableName
-	NewName   TableName
-	IfExists  bool
-	TableSpec *TableSpec
+	Action               string
+	Table                TableName
+	NewName              TableName
+	IfExists             bool
+	TableSpec            *TableSpec
+	PartitionName        ColIdent
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: I was hoping we'd put these things under an *AlterSpec:

```go
type DDL struct {
  Action string
  Table TableName
  NewName TableName
  IfExists bool
  TableSpec *TableSpec
  AlterSpec *AlterSpec
}

type AlterSpec struct {
  PartitionName ColIdent
  PartitionDefinitions []*PartitionDefinition
}
```

Not critical but is more future proof, and we can keep DDL simple. Put all the formatting and such in Alter.

File: go/vt/sqlparser/sql.y [hunk 2]
```diff
@@ -2331,6 +2370,7 @@ reserved_keyword:
 | LOCALTIMESTAMP
 | LOCK
 | MATCH
+| MAXVALUE
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: Did you try these 2 values in `non_reserved_keyword` first? Only put them in `reserved_keyword` if they cause a conflict, otherwise we greatly prefer non_reserved. Just checking.

**Comment by bbeaudreault** [comment]
a few comments on style/organizaiton of the grammar/ast. otherwise looks good!

File: go/vt/sqlparser/sql.y (hunk 2 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: i was following the guidance of:
https://dev.mysql.com/doc/refman/5.7/en/keywords.html

moving MAXVALUE causes 1 shift/reduce, but moving PARTITION to non_reserved produces no conflicts, compiles, and passes tests.

File: go/vt/sqlparser/ast.go (hunk 1 above)
Comment (demmer) [inline go/vt/sqlparser/ast.go]: I agree this would be cleaner and more consistent with create table.

File: go/vt/sqlparser/sql.y [hunk 3]
```diff
@@ -909,6 +914,38 @@ alter_statement:
...
+  }
+| partition_definitions ',' partition_definition
+  {
+    $$ = append($1, $3)
+  }
+
+partition_definition:
+  PARTITION sql_id VALUES sql_id sql_id openb value_expression closeb
+  {
+    if $4.Lowered() != "less" || $5.Lowered() != "than" {
```
Comment (demmer) [inline go/vt/sqlparser/sql.y]: It seems cleaner to define keywords for "less" and "than".

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 4]
```diff
@@ -358,7 +358,15 @@ func (qre *QueryExecutor) checkAccess(authorized *tableacl.ACLResult, tableName
 }
 
 func (qre *QueryExecutor) execDDL() (*sqltypes.Result, error) {
-	ddlPlan := planbuilder.DDLParse(qre.query)
+	sql := qre.query
+	var err error
+	if qre.plan.FullQuery != nil {
```
Comment (demmer) [inline go/vt/vttablet/tabletserver/query_executor.go]: Why is this change necessary -- it seems like you should be able to pass through the original query unmodified rather than sending the generated full query, no?

Comment (demmer) [inline go/vt/vttablet/tabletserver/query_executor.go]: Oh sorry ignore this comment. The whole point of this is bindvar support.

File: go/vt/sqlparser/ast.go (hunk 1 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: let's keep it like this for now. i think the more future-proof thing to do, which i don't want to spend time on right now, is to:
* make DDL an interface
* make Alter a type (along with Create, Drop, Rename, etc)
* on the Alter type, have a PartitionSpec that can be reused betweent the Create+Alter types

File: go/vt/sqlparser/sql.y (hunk 2 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: There are docs on `reserved_keyword` and `non_reserved_keyword` -- those should be used over that keywords.html link.  If the docs are not clear enough feel free to fix as part of this PR.

Our reserved_keyword and non_reserved_keyword are not meant to mirror mysql -- it's for the sake of conflicts in our grammar. In many cases that may be in line with mysql, since we have some similar constructs. But ours is also a much simpler grammar and so less things should be reserved.

File: go/vt/sqlparser/ast.go (hunk 1 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: It should be a relatively simple move around of a couple lines of code to make this new construct be in line with `*TableSpec`. Not sure why we can't do that just because in the future we may or may not want to make legit interfaces for everything.  I think we should start with separate fields for each in the DDL and if it becomes too much down the line we can split it up, and if not then not.

Comment (sougou) [inline go/vt/sqlparser/ast.go]: Sounds like a good idea. I think it should be a trivial change: on line 919 of sql.y, build an AlterSpec that combines the two variables. Or, maybe I'm missing something.

I agree with deferring the interface. That can always be added when needed.

File: go/vt/sqlparser/sql.y (hunk 2 above)
Comment (sougou) [inline go/vt/sqlparser/sql.y]: I've previously misled people by pointing them at the mysql doc. It makes sense to follow the guidelines set by @bbeaudreault because we are a simplified grammar.

But make sure no conflicts are introduced in the grammar :).

**Comment by sougou** [comment]
LGTM
